src/storage/mvcc/txn.rs

```diff
@@ -1032,6 +1039,48 @@ impl<S: Snapshot, P: PdClient + 'static> MvccTxn<S, P> {
         }
     }
 
+    /// Check the status of a secondary (optimistic) lock.
+    ///
+    /// It checks whether the given secondary lock exists. If the lock exists,
+    /// the lock information is returned. Otherwise, it searches the write CF
+    /// for the commit record of the lock and returns the commit timestamp
+    /// (0 if the lock is not committed).
+    ///
+    /// If the lock does not exist or is a pessimistic lock, to prevent the
+    /// status being changed, a rollback may be written and this rollback
+    /// needs to be protected.
+    pub fn check_secondary_lock(
+        &mut self,
+        key: &Key,
+        start_ts: TimeStamp,
+    ) -> Result<(SecondaryLockStatus, Option<ReleasedLock>)> {
+        let mut released_lock = None;
+        let status = match self.reader.load_lock(&key)? {
+            Some(lock) if lock.ts == start_ts => {
+                if lock.lock_type == LockType::Pessimistic {
+                    released_lock = self.unlock_key(key.clone(), true);
+                    SecondaryLockStatus::RolledBack
+                } else {
+                    SecondaryLockStatus::Locked(lock)
+                }
+            }
+            _ => match self.reader.get_txn_commit_info(&key, start_ts)? {
+                Some((commit_ts, write_type)) if write_type != WriteType::Rollback => {
+                    SecondaryLockStatus::Committed(commit_ts)
+                }
+                _ => SecondaryLockStatus::RolledBack,
+            },
+        };
+        // We must protect this rollback in case this rollback is collapsed and a stale
+        // acquire_pessimistic_lock and prewrite succeed again.
+        let write = Write::new_rollback(self.start_ts, true);
+        self.put_write(key.clone(), self.start_ts, write.as_ref().to_bytes());
```

Comment:
This rollback write should be done checking the `status` variable?

Reply:
Ah, sorry. I delete `need_rollback` by mistake when reorganizing the code...
We shoudn't write rollback when the write record exists.

Reply:
I update this PR and add one more test for it.